src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs

```diff
@@ -0,0 +1,1029 @@
+            }
+
+            reader.ThrowIfNotEmpty();
+            iv = iv.Slice(0, bytesWritten);
+        }
+
+        private static unsafe Rfc2898DeriveBytes OpenPbkdf2(
+            ReadOnlySpan<byte> password,
+            ReadOnlyMemory<byte>? parameters,
+            out byte? requestedKeyLength)
+        {
+            if (!parameters.HasValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            Pbkdf2Params pbkdf2Params =
+                AsnSerializer.Deserialize<Pbkdf2Params>(parameters.Value, AsnEncodingRules.BER);
+
+            // No OtherSource is defined in RFC 2898 or RFC 8018, so whatever
+            // algorithm was requested isn't one we know.
+            if (pbkdf2Params.Salt.OtherSource != null)
+            {
+                throw new CryptographicException(
+                    SR.Format(
+                        SR.Cryptography_UnknownAlgorithmIdentifier,
+                        pbkdf2Params.Salt.OtherSource.Value.Algorithm));
+            }
+
+            if (pbkdf2Params.Salt.Specified == null)
+            {
+                Debug.Fail($"No Specified Salt value is present, indicating a new choice was unhandled");
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            HashAlgorithmName prf;
+
+            switch (pbkdf2Params.Prf.Algorithm.Value)
+            {
+                case Oids.HmacWithSha1:
+                    prf = HashAlgorithmName.SHA1;
+                    break;
+                case Oids.HmacWithSha256:
+                    prf = HashAlgorithmName.SHA256;
+                    break;
+                case Oids.HmacWithSha384:
+                    prf = HashAlgorithmName.SHA384;
+                    break;
+                case Oids.HmacWithSha512:
+                    prf = HashAlgorithmName.SHA512;
+                    break;
+                default:
+                    throw new CryptographicException(
+                        SR.Format(
+                            SR.Cryptography_UnknownAlgorithmIdentifier,
+                            pbkdf2Params.Prf.Algorithm));
+            }
+
+            // All of the PRFs that we know about have NULL parameters, so check that now that we know
+            // it's not just that we don't know the algorithm.
+
+            if (!pbkdf2Params.Prf.HasNullEquivalentParameters())
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            int iterationCount = ClampIterationCount(pbkdf2Params.IterationCount);
+            ReadOnlyMemory<byte> saltMemory = pbkdf2Params.Salt.Specified.Value;
+
+            byte[] tmpPassword = new byte[password.Length];
+            byte[] tmpSalt = new byte[saltMemory.Length];
+
+            fixed (byte* tmpPasswordPtr = tmpPassword)
+            fixed (byte* tmpSaltPtr = tmpSalt)
```

Comment:
Why do these need to be pinned?

Reply:
(If it's to be able to clear them and know that no copies have been left in memory, it'd be worth adding a comment to that effect.)

Reply:
If that's the case, though, how does that work once these method returns?  Doesn't the Rfc2898DeriveBytes instance store a reference to the arrays?

Reply:
Some of the pins are proper pin-then-fill; some of them are "we don't have API to allow that" best-efforts.  Since compaction's "shadows" are problematic there's a feature request to coreclr to make compaction clear (and then most of the pinning in this PR can be removed).

Reply:
This one seems particularly useless, though.  We're pinning only briefing and then handing out the arrays inside of the Rfc2898DeriveBytes instance, right?

Reply:
@stephentoub - array is cleared below (this might have got updated since when you wrote the comment)